src/app/modules/item/services/threads.service.ts

```diff
@@ -44,37 +98,115 @@ interface TokenData {
+  private incomingMessages$: Observable<ThreadMessage[]>;
+  private subscriptions: Subscription[] = [];
 
-  constructor() {
+  constructor(
+    private activityLogService: ActivityLogService,
+  ) {
     if (!appConfig.forumServerUrl) throw new Error('cannot instantiate threads service when no forum server url');
     this.socket = webSocket(appConfig.forumServerUrl);
-    this.messages$ = this.socket.pipe(
-      decodeSnakeCase(D.array(threadEventDecoder)),
-      scan((oldEvents, newEvents) => [ ...oldEvents, ...newEvents ]),
+
+    this.incomingMessages$ = this.socket.pipe(
+      decodeSnakeCase(D.array(threadMessageDecoder)),
       catchError(() => EMPTY), // ignore undecoded messages
     );
+
+    this.newEvents$ = this.incomingMessages$.pipe(map(messages => messages.filter(isThreadEventMessage)));
+
+    this.events$ = this.clearEvents$.pipe(
+      startWith(undefined),
+      switchMap(() => this.newEvents$.pipe(scan((oldEvents, newEvents) => [ ...oldEvents, ...newEvents ]))),
+      map(messages => messages.sort((a, b) => a.time.valueOf() - b.time.valueOf())), // sort by date ascending
+    );
+
+    this.status$ = this.incomingMessages$.pipe(
+      map(messages => messages.find(isThreadStatusMessage)?.status),
+      filter(isNotUndefined),
+      distinctUntilChanged(),
+      shareReplay(1),
+    );
+
+    this.subscriptions.push(
+      // Every time we receive an event related to thread status, refetch the status.
+      this.newEvents$.pipe(
+        map(events => events.find(event => event.eventType === 'thread_opened' || event.eventType === 'thread_closed')),
+        filter(isNotUndefined),
+      ).subscribe(() => this.getStatus()),
+    );
+  }
+
+  ngOnDestroy(): void {
+    this.socket.complete();
+    this.subscriptions.forEach(subscription => subscription.unsubscribe());
   }
 
   init(tokenData: TokenData): void {
     this.tokenData = tokenData;
+    this.clearEvents$.next(); // when re-initializing the service, clear event list.
+    this.getStatus();
   }
 
-  send(action: Action): void {
+  private send(action: ThreadAction): void {
     if (!this.tokenData) throw new Error('service must be initialized');
     this.socket.next({
       token: this.tokenData,
-      action,
+      ...action,
+    });
+  }
+
+  open(): void {
+    if (!this.tokenData) throw new Error('cannot open thread without token data');
+
+    const { participantId, userId, itemId } = this.tokenData;
+    const watchedGroupId = participantId === userId ? undefined : participantId;
+
+    this.clearEvents$.next();
+    this.activityLogService.getActivityLog(itemId, watchedGroupId).subscribe({
+      next: history => this.send({ action: 'open-thread', history }),
+      error: () => { /* FIXME: What to do? */ }
     });
   }
 
+  close(): void {
+    this.send({ action: 'close-thread' });
+  }
+
+  follow(): void {
+    this.send({ action: 'follow' });
+  }
+
+  unfollow(): void {
+    this.send({ action: 'unfollow' });
```

Comment:
`unfollow` seems to be called from multiple place. Probably no need to send several of them. 
A "status of the thread" should probably be known/stored in the service so that multiple follow and unfollow are not sent. (with a `throw`  if it is really supposed to be impossible)

Reply:
The fact that a user follows or not does not depend on the thread status anymore, that means: at task init we follow, at task destroy (or window unload) we unfollow.